### ISSUE TITLE
Add/replace TA manually in Normal world

### ISSUE THREAD

**Issue by B1nd00** [OP-TEE/optee_os#5481]
As per the documents, TEE_Supplicant looks for TA in lib/optee_armtz which is in the normal world linux file. How do I add/replace TA manually instead of building the entire thing again?
I am building my TA on the side but want to replace it in the Linux Normal world without building entire qemu/optee as I am making changes often. Can anyone please answer?

**Comment by jforissier** [comment]
Hi @B1nd00,

One way is to use the host mount feature:

```
make -j8 QEMU_VIRTFS_AUTOMOUNT=y run
```
At this point the root of the OP-TEE project is available from QEMU as `/mnt/host` so you can copy your TA from QEMU. For example, with the "Hello world" TA:
```
# cp /mnt/host/out-br/build/optee_examples_ext-1.0/hello_world/ta/out/8aaaf200-2450-11e4-abe2-0002a5d5c51b.ta /lib/optee_armtz/
```

Another option would be to configure and use SSH but that would be much more complicated.

**Comment by B1nd00** [comment]
> Hi @B1nd00,
> 
> One way is to use the host mount feature:
> 
> ```
> make -j8 QEMU_VIRTFS_AUTOMOUNT=y run
> ```
> 
> At this point the root of the OP-TEE project is available from QEMU as `/mnt/host` so you can copy your TA from QEMU. For example, with the "Hello world" TA:
> 
> ```
> # cp /mnt/host/out-br/build/optee_examples_ext-1.0/hello_world/ta/out/8aaaf200-2450-11e4-abe2-0002a5d5c51b.ta /lib/optee_armtz/
> ```
I tried this but the replaced TA is not being recognized. Any idea why? I deleted it then copied the new one to the location and now I cant make a connection. My session isn't even opening

**Comment by DevendraDevadiga** [comment]
OP-TEE OS hands-on for beginners: https://youtu.be/m8WlSanGDwI

**Comment by DevendraDevadiga** [comment]
If you ran the optee_helloworld test before and then copied, may be previously loaded .ta only it's using.

After copy to /lib/optee_armz try to reboot and test.

And please confirm in out-br modified time of your .ta, whether with your changes build or not before copy.


Below are useful links to get hands-on on OP-TEE:
https://youtu.be/m8WlSanGDwI
https://youtu.be/FthDUSW3yQY
https://github.com/DevendraDevadiga/Early-TA
https://github.com/DevendraDevadiga/optee_qemu_armv8a_prebuilt_binaries

**Comment by LXDeathrattle** [comment]
Thank you for being so supportive @DevendraDevadiga.

I met the same problem with @B1nd00 when using Ubuntu18+optee on rpi3.

I followed the instructions of https://optee.readthedocs.io/en/latest/building/gits/optee_examples/optee_examples.html, built only [optee_examples](https://github.com/linaro-swg/optee_examples).

If I don't change the uuid, CAs work. But errors will occur if I replace the TAs in /lib/optee_armtz/ with my own compiled TAs.

The error is:

> TEEC_Opensession failed with code 0xffff000f origin 0x3

If I change the uuid of CA and TA, the error code is the same. I rebooted the system.

Need I to sign TAs? I think TAs had signed during compiling. The key didn't change.

I compiled TAs with the newest version of optee_os but the optee on rpi3 is an old version. I don't know if it's the reason.
